src/circuit/gadget/utilities/cond_swap.rs

```diff
@@ -0,0 +1,301 @@
+                        .value
+                        .zip(b.value)
+                        .zip(swap)
+                        .map(|((a, b), swap)| if swap { b } else { a });
+                    let a_swapped_cell = region.assign_advice(
+                        || "a_swapped",
+                        config.a_swapped,
+                        0,
+                        || a_swapped.ok_or(Error::SynthesisError),
+                    )?;
+                    CellValue {
+                        cell: a_swapped_cell,
+                        value: a_swapped,
+                    }
+                };
+
+                // Conditionally swap b
+                let b_swapped = {
+                    let b_swapped = a
+                        .value
+                        .zip(b.value)
+                        .zip(swap)
+                        .map(|((a, b), swap)| if swap { a } else { b });
+                    let b_swapped_cell = region.assign_advice(
+                        || "b_swapped",
+                        config.b_swapped,
+                        0,
+                        || b_swapped.ok_or(Error::SynthesisError),
+                    )?;
+                    CellValue {
+                        cell: b_swapped_cell,
+                        value: b_swapped,
+                    }
+                };
+
+                // Return swapped pair
+                Ok((a_swapped, b_swapped))
+            },
+        )
+    }
+}
+
+impl<F: FieldExt> CondSwapChip<F> {
+    /// Configures this chip for use in a circuit.
+    ///
+    /// # Side-effects
+    ///
+    /// `advices[0]` will be equality-enabled.
+    pub fn configure(
+        meta: &mut ConstraintSystem<F>,
+        advices: [Column<Advice>; 5],
+    ) -> CondSwapConfig {
+        let a = advices[0];
+        // Only column a is used in an equality constraint directly by this chip.
+        meta.enable_equality(a.into());
+
+        let q_swap = meta.selector();
+
+        let config = CondSwapConfig {
+            q_swap,
+            a,
+            b: advices[1],
+            a_swapped: advices[2],
+            b_swapped: advices[3],
+            swap: advices[4],
+        };
+
+        // TODO: optimise shape of gate for Merkle path validation
+
+        meta.create_gate("a' = b ⋅ swap + a ⋅ (1-swap)", |meta| {
+            let q_swap = meta.query_selector(q_swap);
+
+            let a = meta.query_advice(config.a, Rotation::cur());
+            let b = meta.query_advice(config.b, Rotation::cur());
+            let a_swapped = meta.query_advice(config.a_swapped, Rotation::cur());
+            let b_swapped = meta.query_advice(config.b_swapped, Rotation::cur());
+            let swap = meta.query_advice(config.swap, Rotation::cur());
+
+            let one = Expression::Constant(F::one());
+
+            // a_swapped - b ⋅ swap - a ⋅ (1-swap) = 0
+            // This checks that `a_swapped` is equal to `y` when `swap` is set,
```

Comment:
Should this comment have `b` rather than `y`?

Reply:
Yes.

Reply:
Fixed in #235.